chap-architecture.tex

```diff
@@ -1746,6 +1746,66 @@ \section{Capability-Aware Instructions}
+
+\insnref{CInvoke} unseals the sealed code and data capabilities
+and places them in \PCC{} and \IDC{} (an architecture-specific
+capability register), with control transferred
+directly to the target code capability.
+A programming-language or concurrent programming-framework runtime might
+arrange that all sealed code capabilities point to a message-passing
+implementation that proceeds to check argument registers or clear other
+registers, switching directly to the target domain via a further
+\insnref{CJR}, or returning to the caller if the message will be delivered
+asynchronously.
+
+Voluntary protection-domain crossing -- i.e., not triggered by an interrupt --
+will typically be modeled as a form of function invocation or message passing
+by the operating system.
+In either case, it is important that function callers/callees, message
+senders/recipients, and the operating system itself, be constructed to protect
+themselves from potential confidentiality or integrity problems arising from
+leaked or improperly consumed general-purpose integer registers or
+capabilities passed across domain transition.
+On invocation, callers will wish to ensure that non-argument registers, as
+well as unused argument registers, are cleared.
+Callees will wish to receive only expected argument registers.
+Similarly, on return, callees will wish to ensure that non-return registers,
+as well as unused return registers, are cleared.
+Likewise, callers will wish to receive back only expected return values.
+In practice, responsibility for this clearing lies with multiple of the
+parties: for example, only the compiler may be aware of which argument
+registers are unused for a particular function, whereas the operating system
+or message-passing routine may be able to clear other registers.
+Work performed by the operating system as a trusted intermediary in a
+reliable way may be usefully depended on by either party in order to prevent
+duplication of effort.
+For example, if the OS the clears non-argument
+registers on call, and non-return registers on return, caller and
+callee can avoid clearing those registers allowing that clearing
+to occur exactly once.
+\jhbnote{I'm tempted to replace OS in this paragraph with something
+  like ``domain-transfer supervisor'' since the intermediary may well
+  live in user space (like the runtime linker).}
+Efficient register clearing instructions (e.g., \insnref{CClear}) can
+also be used to substantially accelerate this process.
+
+In CHERI, the semantics of secure message passing or invocation are software defined, and we anticipate that different operating-system and programming-language security models might handle these, and other behaviors, in different ways.
+For example, in our prototype CheriBSD implementation, a userspace supervisor maintains a ``trusted stack'' onto which values are pushed during invocation, and from which values are popped on return.
```

Comment:
I think this is a libcheri thing and should be dropped.

Reply:
It is, though I had just copied this section over with little change from the MIPS chapter.  There are several more libcheri references that need dropping.  For now I just wanted to extricate this content from the MIPS chapter so it wasn't lost.

Reply:
Doing this in a later pass seems fine.